### ISSUE TITLE
Add discrete event visualization

### ISSUE THREAD

**Issue by scottbell** [nasa/openmct#7936]
**Is your feature request related to a problem? Please describe.**
Many telemetry objects may be discrete events. For example, when a command is issued, or a warning is raised. These currently cannot be visualized with data in plots.

**Describe the solution you'd like**
A single vertical line per discrete event, plotted along with numerical telemetry data. Some ideas include:
1. Creating a new widget in the TimeLine object to display these events.
2. Adding discrete event telemetry sources to plots.

**Comment by scottbell** [comment]
I've got events plotting:

https://github.com/user-attachments/assets/86b48cd2-09e8-4eb0-8f79-4b68ae6c5a90

**Comment by mudinthewater** [comment]
@scottbell If you're lookin for a way to categorize Events, the AMMOS repo does a pretty decent job with its EVRs views: https://github.com/NASA-AMMOS/openmct-mcws/tree/main/src/evrView

![image](https://github.com/user-attachments/assets/b41b0955-c736-4011-870d-1c2c30097289)

Then on the view side everything has a configurable color: 

![image](https://github.com/user-attachments/assets/56386b1b-f31c-4095-b569-ca538e8d8650)

For viewing, I like what you're doing with the vertical events - you might want to consider a compact view like this too: 

![image](https://github.com/user-attachments/assets/77107cac-ffe5-45e2-9dfd-b7c871f8927e)

Basically what that view does is puts a line for each EVR level and then a point for each instance of the EVR level on the line so you can see them all in a much more compact way. On mouseover of the point it also populates a popup table that queries the evrs +/- a configured amount of time so you can see the actual record.

**Comment by mudinthewater** [comment]
If you wanted to simulate that last view in MCT, what you'd do is assign a numerical value to each EVR level, e.g.
FATAL = 6, WARN_HI = 5, WARN_LO = 4, ACT_HI = 3, ACT_LO =2, CMD =1, DIAG=0. Then each time an EVR appears that line would get a value at that timestamp. The value is always the numeric value above, and the plot would be set up as an enumeration overlay plot with each level being dragged on it as a telemetry point. Just a slightly more compact way of showing the same information that I think you're lookin for.

**Comment by scottbell** [comment]
These are excellent suggestions @mudinthewater! I'll chat with @charlesh88 and see what we can incorporate into the design.

**Comment by charlesh88** [comment]
### In progress dev notes:
#### Questions
- [ ] Why do we need this structure in EventTimelineView.vue? Can we do this with a single level div and no `u-contents`? 
```
<div ref="events" class="c-events-tsv">
<div ref="eventsHolder" class="c-events-tsv__contents u-contents"></div>
</div>
```
#### Todos
@charlesh88 
- [x] Marcus Bains line style refinements.
- [x] Remove extraneous padding at bottom of plot views when in TS.
- [ ] Convert approach in EventTimelineView.vue to not set an explicit height.
- [ ] Better styling for event and overlay lines.
- [ ] Improved styling, text ellipsis and vertical text for Plan swimlanes.
- [ ] Investigate better hit area approach to event lines.
@scottbell 
- [ ] Attach `is-selected` when an event line is selected; remove on click-outside event.

**Comment by scottbell** [comment]
To test timeline changes
1. `git checkout branch 7936-add-discrete-event-visualization` in openmct
    1. run `npm install && npm link`
2. `git checkout branch 7936-add-discrete-event-visualization` in openmct-yamcs
    1. run `npm install && npm link openmct`
3. `git clone https://github.com/yamcs/quickstart`
    1. run `mvn yamcs:run` in one console
    2. run `./simulator` in another
4. In openmt-yamcs, run `npm run make-example-events`
5. open http://localhost:9000
    1. create a new "Time Strip"
    2. ensure you see sub-objects under myproject -> Commands and myproject -> Events
    3. Test Events
            1. drop "Pressure Monitor" and "Pressure Module" events on the timeline object
            2. change the time conductor to real-time for the last 10 minutes, then back to fixed time
            3. you should see a few events
            4. click on an event, and look at in the inspector. under the event tab, it should show details
            5. hover over the event for a bit, it should show the name
            6. click the button next to title of the swimlane to expand, ensure it enables/disables properly
            7. ensure their colors (though faint) are there. charles will be adjusting these as they're a bit wonky now
    4. Test Commands
        1. Navigate to http://localhost:8090 (YAMCS controls)
        2. Click "Commanding" on the left
        3. Send a few commands (Reboot and VoltageOn for example)
        4. drop "default" and "payload" command sources on the timeline object
        5. change the time conductor to real-time for the last 10 minutes, then back to fixed time
        6. you should see a few commands
            1. click on a command event, and look at in the inspector. under the event tab, it should show details
            2. hover over the command event for a bit, it should show the name
            3. click the button next to title of the swimlane to expand, ensure it enables/disables properly

**Comment by scottbell** [comment]
@charlesh88
> Why do we need this structure in EventTimelineView.vue? Can we do this with a single level div and no u-contents?
Yes, I've removed it and condensed the `<div>`

> Rather than calculating heights on individual elements like event lines, I think we should shift to a world where we set height on the whole swimlane, and then elements inside, like event lines and holders, just set their heights via absolute position offsets or percentages. This will be more performant, and will be better when users can set swimlane heights at time edit.

That sounds good.

>How can I get severity with events locally?
See [testing instructions](https://github.com/nasa/openmct/issues/7936#issuecomment-2545081583) I added.

> Extended event lines don't line up on refresh, must nav away and back to get them to align. Is that an event generator issue, or something else?
Yeah, it's a bug. I'll fix.

> Attach is-selected when an event line is selected; remove on click-outside event.
I can do the former, but the latter (click-outside event) isn't being explicitly done. Other elements (namely the object itself) are generating context events which trigger the inspector

>The event lines need a holder that I can style with a slightly darker background and subtle stroke. This is the same as the plot area, and the left edge must align with the calculated plot margin.

There's an event wrapper we're generating called `wrapper-${utc-date}`. Would that work?

> Hovering on an event line should also set a class or do something to that related extended line as well.
can do

**Comment by charlesh88** [comment]
@scottbell Many updates pushed, see commit messages for detailed notes on changes.

**Comment by charlesh88** [comment]
@scottbell Have pushed changes. Change details in commit notes, but briefly: 
- Event coloring should now be all good across the board. TODO on this is to color the tooltip with the event color.
- Hovers now include transition effects for a better scrubbing experience.
- Styling implemented for `s-selected` on the event lines. 
- Extended lines now get a CSS class of `--hilite` when their related non-extended line is hovered. I need you to also apply that class for an extended line when its non-extended line is also selected, and have aded to Todos above.
- Event Generator now uses a more random delay on event generation time, and sets a severity less frequently.

**Comment by scottbell** [comment]
@mudinthewater thanks again for the suggestion. I added the minimum severity child objects to each event source in YAMCS
<img width="1419" alt="image" src="https://github.com/user-attachments/assets/b86f2f10-87a4-4532-aa6a-7826f2929094" />

**Comment by scottbell** [comment]
Here's the latest and greatest:

https://github.com/user-attachments/assets/3f8500a3-071c-47c2-9bfe-6a354e5de854


https://github.com/user-attachments/assets/9796eedc-5729-42ad-9e16-80a0a7f9be34

**Comment by charlesh88** [comment]
Looks great!

**Comment by charlesh88** [comment]
Pushed significant updates to the Time Strip imagery view:
1. Refactored code and CSS ahead of user-settable swimlane heights. Image thumbs no longer set height and width in code, but instead dynamically size to the height of the container.
2. Imagery strip now uses styles and hover behavior consistent with the Events/Commands view.
3. Copied code from EventTimelineView.vue that gets and sets the imagery container left edge, to properly align the view in time.
![Screenshot 2024-12-19 at 12 46 09 PM](https://github.com/user-attachments/assets/05ff75a3-d677-457e-b6bb-abafcc67be10)
![Screenshot 2024-12-19 at 12 45 36 PM](https://github.com/user-attachments/assets/340d8bf3-7ed5-4f66-9f17-41255d9dd59e)
![Screenshot 2024-12-19 at 12 45 55 PM](https://github.com/user-attachments/assets/6146985a-74e2-4de2-b1db-f59ccf256562)

**Comment by charlesh88** [comment]
> Drag event telemetry into a time strip
> Create an overlay plot with at least 2 telemetry endpoints, one on y-axis 1 and one on y-axis 3 (right hand side).
> Drag an overlay plot in to the time strip
> Observe that the Extended event lines are drawn incorrectly (not handling the right-offset as expected)

This appears to be fixed in a current deployment on our test server.

<img width="1189" alt="Image" src="https://github.com/user-attachments/assets/404a7b28-fffc-441d-a2c4-849e6bae5150" />

<img width="1189" alt="Image" src="https://github.com/user-attachments/assets/9e18ee44-8c26-4196-8cd3-aa1193d86599" />

**Comment by akhenry** [comment]
# Testing notes
* Create a timeline view
* Add a plan
* Add images
* Add telemetry
* Confirm that everything renders correctly and no regressions have been introduced
* Add events
    * Confirm that event color reflects severity. Log into Yamcs and create events of different severity if they don't already exist.
    * Hover over event to see summary
    * Click on an event
        * Event details shown in inspector
* Add commands
    * Confirm that commands are shown as vertical lines
    * Hover over command to see name
    * Click on command to see details in inspector
* Click on expand button next to events and commands to extend vertical lines the full height of the time strip view.

**Comment by jvigliotta** [comment]
Most everything looked good, except the following:
Weird overlap issue (dragged events in and it's overlapped the plan):
<img width="1096" height="683" alt="Image" src="https://github.com/user-attachments/assets/bcb66db2-b39f-4972-8a08-045e0bfa28d7" />

**Comment by akhenry** [comment]
@jvigliotta Believe that this is reported (and fixed) as https://github.com/nasa/openmct/issues/8169

**Comment by akhenry** [comment]
Yamcs events are correctly filtered by source for historical events, but not for real-time events. eg.

![Image](https://github.com/user-attachments/assets/c21127b5-5164-4b65-9142-00719e016dee)

Notice in the above the higher density of events to the right of the time strip, vs older events to the left. This is because events arising in real-time are not being filtered by source.

**Comment by akhenry** [comment]
Reopening because above comment needs to be addressed before we can call this done.

**Comment by akhenry** [comment]
Verified fixed
